106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,534 @@
+```
+
+* from **RebalanceOnScaleUp**:
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is still running or another one was requested while the first one ended.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, so the current rebalancing scale up is stopped (and queued) and a new rebalancing scale down is started. The rebalancing scale up will be postponed.
+  * **RebalanceOnImbalance**: if a ConfigMap related to goal violation was detected. It will run once the queued scale down and scale up is completed.
+  * **Idle**: if a scale up operation was requested, it was executed and completed successfully/failed or a full rebalance was asked due to an anomaly but since the scale-down rebalance is done, we can ignore the anomalies assuming they are fixed by the rebalance. In case, they are not fixed, Cruise Control will detect them again and a new rebalance would be requested.
+
+```mermaid
+sequenceDiagram
+  autonumber
+  RebalanceOnImbalance->>RebalanceOnScaleDown: requested scale down
+  RebalanceOnImbalance->>RebalanceOnScaleUp: requested scale up
+  RebalanceOnImbalance->>RebalanceOnImbalance: new full rebalance requested
+  critical queued scale up and skew rebalance
+    RebalanceOnImbalance->>RebalanceOnScaleUp: execute scale up
+    RebalanceOnScaleUp-->>RebalanceOnImbalance: then full rebalance
+  end
+  critical queued scale down and skew rebalance
+    RebalanceOnImbalance->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnImbalance: then full rebalance
+  end
+  critical queued scale down and scale up and skew rebalance
+    RebalanceOnImbalance->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnImbalance: then full rebalance
+  end
+  RebalanceOnImbalance->>Idle: scale up complete with no queued items
+```
+* from **RebalanceOnImbalance**:
+  * **RebalanceOnImbalance**: if another goal violation was detected while the first one ended. If a scale down and scale up is also queued up then they will execute first.
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is queued and will run if there is no other rebalancing scale down in queue. If a rebalancing scale down is in queue then it will be executed first.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, it will run once the full rebalance is completed
+  * **Idle**: if full rebalance was requested, it was executed and completed successfully or failed.
+
+On each reconciliation, the following process will be used:
+
+```mermaid
+flowchart TB
+  A[KafkaClusterCreator] --creates--> B[KafkaCluster]
+  B -- calls --> D[KafkaAutoRebalancingReconciler.reconcile]
+  D -- check for ConfigMap with goal-violation prefix --> E{if config map present?}
+  D -- if rebalance in progress --> F[moves a;; unfixed anomalies in list to fixed anomalies list]
```

Comment:
What is are these "unfixed" and "fixed anomalies list"(s)? This is the first time you have mentioned them?

Reply:
```suggestion
  D -- if rebalance in progress --> F[moves unfixed anomalies in list to fixed anomalies list]
```